### ISSUE TITLE
Proposed updated release process

### ISSUE THREAD

**Issue by DRMacIver** [HypothesisWorks/hypothesis-python#784]
As I'm sure everyone has noticed by this point, change log wrangling is a *really annoying* part of our current pull request process, makes everyone sad, and slows down the whole process of development. I'd say this is probably our highest priority process problem right now.

I'd like to propose the following fix to it for consideration:

We stop manually updating docs/changes.rst and move to an automated build system for it.

Instead we create a `RELEASE.rst` file at the top level. A pull request for release must create such a file. It contains the proposed changelog entry and some metadata saying what sort of change it is (major/minor/patch).

When the deploy process runs, it looks for that file, figures out from the metadata and the current date what the title should be in the changelog, inserts it into changes.rst, commits those changes, and *then* does the usual tag and deploy malarky.

And thus, no more having to update the version number and date manually in your pull requests 🎉

The only things we can currently do that this prevents are that we can't merge into master until the previous deploy has completed, and we can't rebase off a version of master that is mid-deploy, because they would create merge conflicts. There are some ways we could work around this, but I think the easiest solutions are:

* Don't do that then
* Create a `release` branch that is (automatically updated to be) the latest release tag and which is always safe to rebase off.

What do people think? I'm happy to be the one to implement this but I thought I'd get some feedback first.

**Comment by DRMacIver** [comment]
I'd forgotten #636 when I wrote this. Obviously this is related, but I think this is a cleaner approach than what we discussed there.

**Comment by Zac-HD** [comment]
I like it!  There was some discussion of changelog management at the PyConAU sprints - some people are very fond of directory-based changelogs, which I think are overkill.  Others like VCS-based changelogs (ie extract it all from git metadata), but that's less friendly to new contributors.  Overall I think this proposed system will work well for us.

Re: robustness... *just don't do that*.  Obviously make sure things attempt the parts of deployment in the right order and fail if something bad happens (eg non-force push, and deploy to PyPI iff that succeeds).  Anything issues that *do* arise can be resolved by the repo owner, and as long as nothing can rewrite git history or cause wrong uploads to PyPI it's all fixable.

**Comment by DRMacIver** [comment]
> directory-based changelogs, which I think are overkill.

I don't think it's necessarily overkill - I thought about doing that for this, and combining them from there, but the main reason I decided against it was that it seemed messier to automate and with more complicated failure modes.

> Others like VCS-based changelogs (ie extract it all from git metadata), but that's less friendly to new contributors.

I really dislike this approach FWIW. I think what you want to include in git is very different from what you want to show to users.

> Anything issues that do arise can be resolved by the repo owner

One of the nice things about this is that it generally "fails safe" - there aren't too many issues that *can* arise, because they all turn into very obvious merge conflicts that you have to resolve before merging.

Those are probably famous last words though.

**Comment by Zac-HD** [comment]
- We do `git rm RELEASE.rst`, but not `git add RELEASE.rst` - so it's still there.
- The `check-rst` job needs to be updated.  We should merge `RELEASE.rst` into the changelog file before this check, and then exclude the release file from it.  (as-is, we get [cross-reference errors](https://travis-ci.org/HypothesisWorks/hypothesis-python/jobs/265390056))

Ping @DRMacIver, and sorry about the number of messages!

**Comment by DRMacIver** [comment]
> We do git rm RELEASE.rst, but not git add RELEASE.rst - so it's still there.

I think you're being confused by the fact that the release process didn't work. git rm updates the index as well as deleting the file, so there's no need to git add it. The file is still there because the release hasn't happened yet!

> The check-rst job needs to be updated. We should merge RELEASE.rst into the changelog file before this check, and then exclude the release file from it. (as-is, we get cross-reference errors)

Ah, that's a good point.

**Comment by Zac-HD** [comment]
> I think you're being confused by the fact that the release process didn't work. 

Ah, that does make more sense.  I'll just rebase against the commit before that merge for now.

**Comment by DRMacIver** [comment]
I'm currently trying to fix it by pushing fixes straight to master, which never goes wrong of course. 😉 Hopefully https://travis-ci.org/HypothesisWorks/hypothesis-python/builds/265453258 should get us a release.

**Comment by DRMacIver** [comment]
Spoilers: It went wrong.

It successfully pushed to pypi but failed in the tagging. I've manually hand cranked a tag that isn't actually the right commit ID but has the same contents, and I've pushed a fix for the problem that caused that.

So the current state is:

* We have successfully released and master is in the state it should be
* It *might* work correctly when we next do a release.

**Comment by Zac-HD** [comment]
Well, that's progress!

I volunteer #790 as the next release to attempt - it's finished, rebased on master, one commit, and shouldn't need more review than a final skim.
